### PR TITLE
Refactor to fix trafficgen UIDs

### PR DESCRIPTION
**Pull request by portante**
The `pbench-trafficgen` script generates unique iterations based on the following command line argument (provided or defaulted):

 * `--traffic-profiles`
 * `--rates`
 * `--traffic-directions`
 * `--frame-sizes`
 * `--num-flows`
 * `--max-loss-pcts`

However, the UIDs for each generated sample was using the following set of data:

 * `-traffic_generator:%traffic_generator%-trial:%trial%`

This can lead to situations where one asks for `pbench-trafficgen` to run multiple iterations, where if the trial numbers don't match the iteration data won't be merged.

This set of modifications ensures that the UIDs used in the generated data match the values used to create each iteration.

One of the behaviors observed previously was that one could see an error like the following (long lines artificially wrapped):

```
Error: values do not match
key: claimed_dev_pairs values: ARRAY(0x559b88b50c20) ARRAY(0x559b88b5f750)
Error: values do not match
key: null_stats values: HASH(0x559b88b5d230) HASH(0x559b88b5df68)
Error: values do not match
key: output_dir values: /var/lib/pbench-agent/\
  trafficgen_RHOSP16.1-RHEL8.2-OVS-OFFLOAD_tg:trex_r:none_fs:1500_nf:128_fm:si_td:bi_ml:0.002_tt:bs__2021-01-13T10:43:04/\
  1-bidirectional-1500B-128flows-0.002pct_drop/sample2 /var/lib/pbench-agent/\
  trafficgen_RHOSP16.1-RHEL8.2-OVS-OFFLOAD_tg:trex_r:none_fs:1500_nf:128_fm:si_td:bi_ml:0.002_tt:bs__2021-01-13T10:43:04/\
  1-bidirectional-1500B-128flows-0.002pct_drop/sample3
Error: values do not match
key: test_dev_pairs values: ARRAY(0x559b88b5d500) ARRAY(0x559b88b5f498)
```

This output was due in part to the UIDs not matching because the `trial` value was different.  But even if we resolve that difference, the `output_dir` key will always be different between samples, causing this error to be emitted.  Further, differences between Perl arrays and hashes were not being handled properly, causing the other errors emitted above.

This change addresses those problems as well.

### CODE REVIEW

**Review by ndokos — CHANGES_REQUESTED** [review]
In the commit message (and the PR description): "This output was due in part to the UIDs not match ..."  should be  " .... not matching ...".

In `process_iteration_samples` there are commented-out debugging print statements: should they be deleted?

**Review by portante — COMMENTED** [review]
From a code walk-through...

File: agent/bench-scripts/tests/pbench-trafficgen/test-22.txt [hunk 1]
```diff
@@ -2119,17 +2119,18 @@ Iteration 3-revunidirectional-74B-1000flows-0.0pct_drop complete (3 of 3), with
...
                   "traffic_profile" : "/var/lib/pbench-agent/trafficgen_test-22_tg[trex-profile]r[5000000]fs[74]nf[1000]fm[si,di,sp,dp,protocol]td[bi,uni,revuni]ml[0.0]tt[os]_2018-09-24T19:10:50/1-bidirectional-74B-1000flows-0.0pct_drop/traffic-profile.json",
                   "trex_profiler_interval" : 3,
-                  "trial" : "1",
+                  "trial" : 1,
                   "trial_mode" : "validation",
                   "trial_primary_output_file" : "binary-search.trial-001.txt",
                   "trial_profiler_file" : "binary-search.trial-001.profiler.txt",
                   "trial_secondary_output_file" : "binary-search.trial-001.extra.txt",
-                  "uid" : "trafficgen-unknown-traffic_generator:%traffic_generator%-trial:%trial%",
+                  "uid" : "trafficgen-unknown-rate:%rate_arg%-traffic_direction:%traffic_direction%",
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-22.txt:2127]: Fixed.

File: agent/bench-scripts/tests/pbench-trafficgen/test-22.txt [hunk 2]
```diff
@@ -1,16 +1,16 @@
 +++ Running test-22 pbench-trafficgen
 tool trigger strings for start: "Starting binary-search" and for stop: "Finished binary-search" are now registered for tool group: "default"
-[warn][1900-01-01T00:00:00.000000] [pbench-trafficgen] ignoring --config="test-22_tg:trex-profile_r:5000000_fs:74_nf:1000_fm:si,di,sp,dp,protocol_td:bi,uni,revuni_ml:0.0_tt:os_" in favor of --rundir="/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-22_1970.01.01T00.00.00"
+[warn][1900-01-01T00:00:00.000000] [pbench-trafficgen] ignoring --config="test-22_tg:trex-profile_r:5000000_fs:74_nf:1000_fm:si,di,sp,dp,protocol_td:bi,uni,revuni_ml:0.0_tt:os" in favor of --rundir="/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-22_1970.01.01T00.00.00"
 Total number of benchmark iterations: 3
-Starting iteration[1-bidirectional-74B-1000flows-0.0pct_drop] (1 of 3)
+Starting iteration[1-100-bidirectional-74B-1000flows-0.0pct_drop] (1 of 3)
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-22.txt:5]: Fixed.

File: agent/bench-scripts/tests/pbench-trafficgen/test-39.txt [hunk 3]
```diff
@@ -2,75 +2,72 @@
 tool trigger strings for start: "Starting binary-search" and for stop: "Finished binary-search" are now registered for tool group: "default"
 found IOMMU option, continuing
 found dpdk-tools, continuing
-[warn][1900-01-01T00:00:00.000000] [pbench-trafficgen] ignoring --config="test-39_tg:trex-profile_r:5000000_fs:74_nf:1000_fm:si,di,sp,dp,protocol_td:bi,uni,revuni_ml:0.0_tt:os_" in favor of --rundir="/var/tmp/pbench-test-bench/pbench-agent"
+[warn][1900-01-01T00:00:00.000000] [pbench-trafficgen] ignoring --config="test-39_tg:trex-profile_r:5000000_fs:74_nf:1000_fm:si,di,sp,dp,protocol_td:bi,uni,revuni_ml:0.0_tt:os" in favor of --rundir="/var/tmp/pbench-test-bench/pbench-agent"
 Total number of benchmark iterations: 3
-Starting iteration[1-bidirectional-74B-1000flows-0.0pct_drop] (1 of 3)
+Starting iteration[1-100-bidirectional-74B-1000flows-0.0pct_drop] (1 of 3)
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-39.txt]: Fixed.

**Review by dbutenhof — APPROVED** [review]
Wow.

I once read a story where one character told a long and implausible story which made a strange assertion, and ended with "Is this true?" and the other character responded "I don't know, but you tell it well".

I think that response is extremely appropriate here.

File: agent/bench-scripts/postprocess/process-iteration-samples [hunk 4]
```diff
@@ -188,12 +195,12 @@ foreach $sample (@json_samples) { # samples 0..N
...
 		foreach my $metric_name (sort keys %{ $sample_hash{$metric_type} }) {
 			for (my $i = 0; $i < scalar @{ $sample_hash{$metric_type}{$metric_name} }; $i++) {
-			my $mapped_uid = get_uid($sample_hash{$metric_type}{$metric_name}[$i]{get_label('uid_label')}, \%{ $sample_hash{$metric_type}{$metric_name}[$i] });
+				my $mapped_uid = get_uid($sample_hash{$metric_type}{$metric_name}[$i]{get_label('uid_label')}, \%{ $sample_hash{$metric_type}{$metric_name}[$i] });
 				if ($mapped_uid) {
 					# either find metric_name element within %iteration with same uid, or create a new one
 					my $iteration_matching_index;
 					if ($iteration{$metric_type}{$metric_name}) {
-						for (my $j = 0; $j < scalar @{$iteration{$metric_type}{$metric_name}}; $j++) {
+						for (my $j = 0; $j < scalar @{ $iteration{$metric_type}{$metric_name} }; $j++) {
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/bench-scripts/postprocess/process-iteration-samples:203]: Seems like the combination is scarier!

File: agent/bench-scripts/postprocess/process-iteration-samples [hunk 5]
```diff
@@ -209,9 +216,13 @@ foreach $sample (@json_samples) { # samples 0..N
 						my %sample = (get_label('uid_label') => $sample_hash{$metric_type}{$metric_name}[$i]{get_label('uid_label')});
 						unshift( @{ $iteration{$metric_type}{$metric_name} }, \%sample);
 						$iteration_matching_index = 0;
+						#print "NEW $mapped_uid -- [$metric_type][$metric_name][$i] ==> [$metric_type][$metric_name][$iteration_matching_index]\n";
+					} else {
+						#print "EXISTING $mapped_uid -- [$metric_type][$metric_name][$i] ==> [$metric_type][$metric_name][$iteration_matching_index]\n";
 					}
 					foreach my $key (sort keys %{$sample_hash{$metric_type}{$metric_name}[$i]}) {
 						my $sample_value = $sample_hash{$metric_type}{$metric_name}[$i]{$key};
+						#print "[$metric_type][$metric_name][$i][$key] = $sample_value \n";
```
Comment (ndokos) [inline agent/bench-scripts/postprocess/process-iteration-samples:225]: Here and above (ll. 219-221): do we want to keep the debugging output?

File: agent/bench-scripts/postprocess/process-iteration-samples [hunk 6]
```diff
@@ -261,6 +282,8 @@ foreach my $metric_name (sort keys %iteration) {
 		for ($i = 0; $i < scalar @{$iteration{$metric_name}{$metric_type}}; $i++) {
 			my $key;
 			foreach $key (sort keys %{$iteration{$metric_name}{$metric_type}[$i]}) {
+				#my $sample_value = $iteration{$metric_name}{$metric_type}[$i]{$key};
+				#print "[$metric_name][$metric_type][$i][$key] = $sample_value \n";
```
Comment (ndokos) [inline agent/bench-scripts/postprocess/process-iteration-samples:286]: Here too: do we want to keep the debugging output?

**Review by ndokos — CHANGES_REQUESTED** [review]

File: agent/bench-scripts/tests/pbench-trafficgen/test-50.txt [hunk 7]
```diff
@@ -0,0 +1,105 @@
++++ Running test-50 pbench-trafficgen
+tool trigger strings for start: "Starting binary-search" and for stop: "Finished binary-search" are now registered for tool group: "default"
+found IOMMU option, continuing
+found dpdk-tools, continuing
+Total number of benchmark iterations: 1
+cp: cannot stat '/var/tmp/pbench-test-bench/pbench-agent/tmp/trex_cfg.yaml': No such file or directory
```
Comment (riya-17) [inline agent/bench-scripts/tests/pbench-trafficgen/test-50.txt:6]: Seems like we are facing something unusual here.

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt [hunk 8]
```diff
@@ -0,0 +1,310 @@
++++ Running test-51 pbench-trafficgen
+tool trigger strings for start: "Starting binary-search" and for stop: "Finished binary-search" are now registered for tool group: "default"
+found IOMMU option, continuing
+found dpdk-tools, continuing
+Total number of benchmark iterations: 4
+cp: cannot stat '/var/tmp/pbench-test-bench/pbench-agent/tmp/trex_cfg.yaml': No such file or directory
+cp: cannot stat 'profile1': No such file or directory
```
Comment (riya-17) [inline agent/bench-scripts/tests/pbench-trafficgen/test-51.txt:7]: Was it supposed to fail?

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+[pbench-tool-trigger]pbench-stop-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/1-default/sample2
+test sample 3 of 3
+[pbench-tool-trigger] starting trigger processing of STDIN using tool group default triggers at /var/tmp/pbench-test-bench/pbench-agent/tools-v1-default/__trigger__
+[pbench-tool-trigger] start-trigger:"Starting binary-search" stop-trigger:"Finished binary-search"
+Starting binary-search
+[pbench-tool-trigger]pbench-start-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/1-default/sample3
+Finished binary-search
+[pbench-tool-trigger]pbench-stop-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/1-default/sample3
+Iteration 1-profile1-0.002pct_drop complete (1 of 4), with 1 pass and 0 failures
+cp: cannot stat 'profile1': No such file or directory
```
Comment (riya-17) [inline agent/bench-scripts/tests/pbench-trafficgen/test-51.txt:31]: was it supposed to behave like this?

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+[pbench-tool-trigger]pbench-stop-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/2-default/sample2
+test sample 3 of 3
+[pbench-tool-trigger] starting trigger processing of STDIN using tool group default triggers at /var/tmp/pbench-test-bench/pbench-agent/tools-v1-default/__trigger__
+[pbench-tool-trigger] start-trigger:"Starting binary-search" stop-trigger:"Finished binary-search"
+Starting binary-search
+[pbench-tool-trigger]pbench-start-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/2-default/sample3
+Finished binary-search
+[pbench-tool-trigger]pbench-stop-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/2-default/sample3
+Iteration 2-profile1-0.001pct_drop complete (2 of 4), with 1 pass and 0 failures
+cp: cannot stat 'profile2': No such file or directory
```
Comment (riya-17) [inline agent/bench-scripts/tests/pbench-trafficgen/test-51.txt:55]: Same here. Seems like copy operations are failing

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+[pbench-tool-trigger]pbench-stop-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/3-default/sample2
+test sample 3 of 3
+[pbench-tool-trigger] starting trigger processing of STDIN using tool group default triggers at /var/tmp/pbench-test-bench/pbench-agent/tools-v1-default/__trigger__
+[pbench-tool-trigger] start-trigger:"Starting binary-search" stop-trigger:"Finished binary-search"
+Starting binary-search
+[pbench-tool-trigger]pbench-start-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/3-default/sample3
+Finished binary-search
+[pbench-tool-trigger]pbench-stop-tools --group=default --dir=/var/tmp/pbench-test-bench/pbench-agent/trafficgen_test-51_tg:trex-profile_pf:profile1,profile2_ml:0.002,0.001_tt:bs_1900-01-01T00:00:00/3-default/sample3
+Iteration 3-profile2-0.002pct_drop complete (3 of 4), with 1 pass and 0 failures
+cp: cannot stat 'profile2': No such file or directory
```
Comment (riya-17) [inline agent/bench-scripts/tests/pbench-trafficgen/test-51.txt:79]: Seems like copy operations are failing

File: agent/bench-scripts/postprocess/process-iteration-samples (hunk 5 above)
Comment (portante) [reply agent/bench-scripts/postprocess/process-iteration-samples]: Removed.

File: agent/bench-scripts/postprocess/process-iteration-samples (hunk 6 above)
Comment (portante) [reply agent/bench-scripts/postprocess/process-iteration-samples]: Removed.

File: agent/bench-scripts/tests/pbench-trafficgen/test-50.txt (hunk 7 above)
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-50.txt]: This was missing a command line option.

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt (hunk 8 above)
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-51.txt]: No, I have added mock profile files to the test.

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt (hunk 9 above)
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-51.txt]: No, I have added mock profile files to the test.

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt (hunk 10 above)
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-51.txt]: Fixed.

File: agent/bench-scripts/tests/pbench-trafficgen/test-51.txt (hunk 11 above)
Comment (portante) [reply agent/bench-scripts/tests/pbench-trafficgen/test-51.txt]: Yes, the profile files were missing; I have added mock profile files to the test.

**Review by dbutenhof — APPROVED** [review]

**Review by npalaska — APPROVED** [review]

**Review by ndokos — APPROVED** [review]

**Pull request merged**